go/pkg/transposeservice/transpose-service.1.md
# transpose-service 1 "" Netspoc "User Manual"

# NAME

transpose-service - Transpose source and destination in service definitions

# SYNOPSIS

transpose-service [options] FILE|DIR [service:]NAME ...

# DESCRIPTION

This program reads a Netspoc configuration and one or more SERVICE name(s).
It transposes the specified services in each file by switching the `user`
keyword between `src` and `dst` attributes in service rules. The
functionality of the service is not changed - the traffic flow remains
the same, only the perspective (which side is considered the `user`) 
is reversed.

For example, if a service has `user = src` and allows traffic from
network A to network B, after transposition it will have `user = dst`
and still allow the same traffic from network A to network B, but now
the destination (network B) is considered the `user` instead of the source.

Changes are done in place, no backup files are created. But only
changed files are touched.

Service names can be specified with or without the `service:` prefix.

## Limitations

A service can only be transposed if:

- It does not use the `foreach` keyword
- It has exactly one rule (not zero, not multiple)
- The rule has both `src` and `dst` attributes defined

# OPTIONS

**-f**, **--file** file
:   Read SERVICE names from file.

**-q**, **--quiet**
:   Don't show changed files.

**-h**, **--help**
:   Print a brief help message and exit.

# EXAMPLES

Transpose service:s1 in Netspoc configuration in directory `netspoc/`:

`transpose-service netspoc/ service:s1`

This would change a service like:

    service:s1 = {
     user = network:n1;
     permit src = user; dst = network:n2; prt = tcp 80;
    }

to:

    service:s1 = {
     user = network:n2;
     permit src = network:n1; dst = user; prt = tcp 80;
    }

Transpose multiple services:

`transpose-service netspoc/ service:s1 service:s2`

Transpose services specified in a file:

`echo 'service:s1 service:s2' > services.txt;
 transpose-service -f services.txt netspoc/`

# COPYRIGHT AND DISCLAIMER

(c) 2025 by Dominik Kunkel, netspoc@drachionix.eu

(c) 2025 by Heinz Knutzen, heinz.knutzen@googlemail.com

http://hknutzen.github.com/Netspoc

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

Changes
{{$NEXT}}

 - export-netvis: export Dualstack-Address of network
   in field Address2 of node.

6.079     2025-03-17 12:53:54+01:00 Europe/Berlin

 - If service has some owner with name "DA_*",
   this service is marked as visible for all owners with name "DA_*"
   in exported data for Netspoc-Web.
 - Command 'print-path' now accepts dual stack objects.

6.078     2025-02-24 14:42:54+01:00 Europe/Berlin

 - New commandline programs export-netvis and print-path have been added.

    These tools output information as JSON to be used in
    a graph visualization.

    export-netvis creates a list of all networks and routers
    of the given topology with information useful for visualizing
    the topology.

    print-path analyzes the given topology and creates a list of
    names of network nodes (routers and networks) between source and
    destination passed via command line.
    To create the list, print-path uses the same path-walk algorithm
    as used in netspoc main program to create ruleset for devices.

 - Aggregate with IPv4 IP 0.0.0.0/0 no longer matches IPv6 addresses.
   Aggregate with IPv6 IP ::/0 no longer matches IPv4 addresses.
 - Fixed split service from auto intf with IPv4, IPv6
 - Command 'export-netspoc' no longer strips comments and ";"
   from description in details of exported service.

6.077     2025-01-27 11:24:26+01:00 Europe/Berlin

 - Allow dual stack service to have protocols icmp and icmpv6 together.
   Generated rules automatically will use icmp for IPv4 and icmpv6 for IPv6.
 - Removed support for separate IPv6 topology in subdirectory 'ipv6/'.
 - Removed options '--ipv6', '-6'.
 - Removed support for NAT of IPv6 objects.
   After changing attributes of IPv6 addresses from 'ip' to 'ip6',
   supporting NAT for IPv6 would need to change NAT attributes
   to 'ip6', 'hidden6', 'identity6' as well.
   This seems to be superflous, since nobody uses NAT for IPv6.
 - Fixed duplicate hosts in print-group from wrong IPv4 network
   in IPv6 aggregate.
 - Fixed panic in print-group from wrong IPv4 network in IPv6 hosts
   generated by 'auto_ipv6_hosts'.
 - Fixed panic in export-netspoc with rule from IPv4 network to
   combined IPv4/v6 auto interface.
 - Changed 'format-netspoc' to print attributes "ip" and "ip6" in
   one line, if object has no more than 3 attributes.

6.076     2025-01-10 09:39:54+01:00 Europe/Berlin

 - Fixed dual stack bridge. IPv4 and IPv6 interfaces were accidently
   processed together.

6.075     2024-12-19 11:53:37+01:00 Europe/Berlin

 - Fix: Attribute 'policy_distribution_point' at dual stack router
   is only used for matching IPv4/IPv6 part of this router.
 - Fixed command 'export-netspoc'.
   Show IPv4 areas for combined zone. Must not accidently overwrite
   IPv4 areas by IPv6 areas, when combined zones have same name.

6.074     2024-12-16 11:52:23+01:00 Europe/Berlin

 - Fixed wrong error on duplicate interface from router in ipv6/ directory.

6.073     2024-12-16 10:16:33+01:00 Europe/Berlin

 - Added support for dual stack objects, having both, IPv4 and IPv6 addresses.
   This simplifies the modeling of a dual stack toplology.
   It is no longer neccessary to model a separate IPv4 and IPv6 topology.
   Rules between dual stack objects will generate ACls for IPv4 and IPv6.
   Rules between dual stack object and pure IPv4 object will silently ignore
   IPv6 address and generate only ACL for IPv4.
   New attributes have been introduced to define dual stack objects:
   - 'ip6' at network, host, interface and aggregate.
   - 'range6' at host.
   - 'unnumbered6' at network and interface.
   - 'negotiated6' at interface.
   Other changes resulting from use of dual stack objects:
   - If dual stack objects are used as border or inclusive border of an area,
     this defines two areas with identical name:
     one in IPv4 topology and one in IPv6 topology.
   - If dual stack objects are used as interfaces of an pathrestriction,
     this also defines two pathrestrictions in IPv4 and IPv6 topology.
     If the second pathrestriction has only one interface or only interfaces
     outside of a loop, it is silently ignored.
   - New attributes 'ipv4_only' and 'ipv6_only' may be used at service or area.
     This will enable only IPv4 or IPv6 part of dual stack objects.
   - The following attributes are applied only to IPv4 part
     if used in dual stack objects:
     'nat', 'bind_nat', 'subnet_of, 'hub', 'spoke'.
   - Command 'export-netspoc' has been changed to show IPv6 address of
     dual stack objects in attribute 'ip6'.
 - Added new attribute 'auto_ipv6_hosts' to automatically generate
   dual stack hosts from pure IPv4 hosts.
   It will generate IPv6 addresses for hosts
   by combining its IPv4 adress with the IPv6 address of its network.
   These attribute values are provided:
   - auto_ipv6_hosts = readable
     network: ip6 = 2001:db8:1:1::/64;
     hosts: ip = 172.17.1.48;
     ==> 2001:db8:1:1:172:17:1:48
   - auto_ipv6_hosts = binary
     network: ip6 = 2001:db8:1:1::/64;
     hosts: ip = 172.17.1.48;
     ==> 2001:db8:1:1::ac11:130
   - auto_ipv6_hosts = none
     No IPv6 address is generated
   This attribute can be used at network, area and host.
 - Fixed command 'cut-netspoc':
   Full path to management_instance is marked now.

6.072     2024-09-05 10:22:13+02:00 Europe/Berlin

 - Fixed command 'remove-from-netspoc'.
   A service is no longer removed accidently, if only some rules of
   a service become empty, but other rules remain.
 - Message "swap objects of 'user' and objects of rules" is shown
   even if no attribute multi_owner is given.
 - Fixed check for NAT network being subnet of some other network.
   If multiple networks have dynamic NAT to same address,
   now a warning is given for each network.
 - Fixed check for useless attribute 'subnet_of'
   where supernet has attribute 'has_subnets'.

6.071     2024-07-29 16:47:58+02:00 Europe/Berlin

 - Fixed generated ACL at router R with managed=local for rule where

go/pkg/transposeservice/transpose-service.go
package transposeservice

/*
=head1 NAME

transpose-service - Transpose one service in netspoc files

=head1 SYNOPSIS

transpose-service [options] FILE|DIR [service:]NAME ...

=head1 DESCRIPTION

This program reads a netspoc configuration and one or more SERVICE name(s). It
transposes the specified service in each file. Keyword user in src/dst is switched
but functionality of the service is not changed. Filechanges are done in place,
no backup files are created. But only changed files are touched.

=head1 OPTIONS

=over 4

=item B<-f> file

Read services from file.

=item B<-q>

Quiet, don't print status messages.

=item B<-h>

Prints a brief help message and exits.

=back

=head1 COPYRIGHT AND DISCLAIMER

(c) 2023 by Dominik Kunkel <netspoc@drachionix.eu>
(c) 2022 by Heinz Knutzen <heinz.knutzen@googlemail.com>

http://hknutzen.github.com/Netspoc

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

import (
	"errors"
	"fmt"
	"os"
	"slices"
	"strings"

	"github.com/hknutzen/Netspoc/go/pkg/ast"
	"github.com/hknutzen/Netspoc/go/pkg/astset"
	"github.com/hknutzen/Netspoc/go/pkg/oslink"
	"github.com/spf13/pflag"
)

type state struct {
	*astset.State
}

func (s *state) readObjects(path string) ([]string, error) {
	bytes, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("Can't %s", err)
	}
	objects := strings.Fields(string(bytes))
	return objects, nil
}

func (s *state) transposeService(name string) error {
	var errMsg string
	found := s.Modify(func(toplevel ast.Toplevel) bool {
		n, ok := toplevel.(*ast.Service)
		if !ok || n.GetName() != name {
			return false
		}
		if n.Foreach {
			errMsg = "Can't transpose service: foreach present."
			return false
		}
		if len(n.Rules) > 1 {
			errMsg = "Can't transpose service: multiple rules present."
			return false
		}
		if len(n.Rules) == 0 {
			errMsg = "Can't transpose service: no rule present."
			return false
		}
		srcElements := n.Rules[0].Src.Elements
		dstElements := n.Rules[0].Dst.Elements
		userElements := n.User.Elements
		srcIsUser := len(srcElements) == 1 && srcElements[0].GetType() == "user"
		dstIsUser := len(dstElements) == 1 && dstElements[0].GetType() == "user"
		if srcIsUser && dstIsUser {
			errMsg = "Can't transpose service: both src and dst reference user."
			return false
		}
		if !(srcIsUser || dstIsUser) {
			errMsg = "Can't transpose service:" +
				" none of src and dst directly reference user."
			return false
		}
		if srcIsUser {
			if hasUserInElements(dstElements) {
				errMsg = "Can't transpose service: dst references nested user."
				return false
			}
			n.Rules[0].Src.Elements = userElements
			n.Rules[0].Dst.Elements = srcElements
			n.User.Elements = dstElements
		} else {
			if hasUserInElements(srcElements) {
				errMsg = "Can't transpose service: src references nested user."
				return false
			}
			n.Rules[0].Src.Elements = dstElements
			n.Rules[0].Dst.Elements = userElements
			n.User.Elements = srcElements
		}
		return true
	})

	if found {
		return nil
	}
	if errMsg == "" {
		errMsg = "Can't find service " + name
	}
	return errors.New(errMsg)
}

func hasUserInElements(l []ast.Element) bool {
	var hasUser func(el ast.Element) bool
	hasUser = func(el ast.Element) bool {
		switch x := el.(type) {
		case *ast.User:
			return true
		case ast.AutoElem:
			return hasUserInElements(x.GetElements())
		case *ast.Intersection:
			return hasUserInElements(x.Elements)
		case *ast.Complement:
			return hasUser(x.Element)
		default:
			return false
		}
	}
	return slices.ContainsFunc(l, hasUser)
}

func Main(d oslink.Data) int {
	fs := pflag.NewFlagSet(d.Args[0], pflag.ContinueOnError)

	// Setup custom usage function.
	fs.Usage = func() {
		fmt.Fprintf(d.Stderr,
			"Usage: %s [options] FILE|DIR [service:]NAME ...\n%s",
			d.Args[0], fs.FlagUsages())
	}

	// Command line flags
	quiet := fs.BoolP("quiet", "q", false, "Don't show changed files")
	fromFile := fs.StringP("file", "f", "", "Read SERVICES from file")
	if err := fs.Parse(d.Args[1:]); err != nil {
		if err == pflag.ErrHelp {
			return 1
		}
		fmt.Fprintf(d.Stderr, "Error: %s\n", err)
		fs.Usage()
		return 1
	}

	s := new(state)

	// Argument processing
	args := fs.Args()
	var services []string
	if *fromFile != "" {
		l, err := s.readObjects(*fromFile)
		if err != nil {
			fmt.Fprintf(d.Stderr, "Error: %s\n", err)
			return 1
		}
		services = l
	} else if len(args) == 0 {
		fs.Usage()
		return 1
	}
	services = append(services, args[1:]...)

	path := args[0]
	var err error
	s.State, err = astset.Read(path)
	if err != nil {
		fmt.Fprintf(d.Stderr, "Error while reading netspoc files: %s\n", err)
		return 1
	}

	for _, service := range services {
		name := service
		if !strings.HasPrefix(service, "service:") {
			name = "service:" + service
		}
		err = s.transposeService(name)
		if err != nil {
			fmt.Fprintf(d.Stderr, "Error: %s\n", err)
			return 1
		}
	}
	s.ShowChanged(d.Stderr, *quiet)
	s.Print()
	return 0
}
